server/src/main/java/org/apache/iotdb/db/query/dataset/AlignByDeviceDataSet.java

```diff
@@ -230,14 +230,29 @@ public boolean hasNextWithoutConstraint() throws IOException {
   protected Set<String> getDeviceMeasurements(PartialPath device) throws IOException {
     try {
       IMNode deviceNode = IoTDB.metaManager.getNodeByPath(device);
-      Set<String> res = new HashSet<>(deviceNode.getChildren().keySet());
+      Set<String> res = new HashSet<>();
       for (IMNode mnode : deviceNode.getChildren().values()) {
-        res.addAll(mnode.getChildren().keySet());
+        IMeasurementSchema measurementSchema = ((MeasurementMNode) mnode).getSchema();
+        if (measurementSchema instanceof VectorMeasurementSchema) {
+          for (String subMeasurement : measurementSchema.getSubMeasurementsList()) {
+            res.add(measurementSchema.getMeasurementId() + "." + subMeasurement);
+          }
+        } else {
+          res.add(mnode.getName());
+        }
```

Comment:
I suggest encapsulate this measurement access in MManager. It's better not use MNode outside the metadata module.

Reply:
You are right. Since it's a special method only used for align by device, I just don't want to make MManager more complex. I will try to fix it.